# HG changeset patch
# User Marco Caliari  <marco.caliari@univr.it>
# Date 1300254331 25200
#      Tue Mar 15 22:45:31 2011 -0700
# Node ID 919cadf334f8995e2cab922ee1a19a8bcd27bd00
# Parent  41d183070c04f3374e80a593a1232412b964026c
Simplify calculation of number of tick labels.  Bug #32692.

diff --git a/src/ChangeLog b/src/ChangeLog
--- a/src/ChangeLog
+++ b/src/ChangeLog
@@ -1,8 +1,13 @@
+2010-03-15  Marco Caliari  <marco.caliari@univr.it>
+
+	* graphics.cc: Simplify calculation of number of tick labels.  Fixes
+	bug #32692.
+
 2011-03-13  Konstantinos Poulios  <logari81@googlemail.com>
 
 	* DLD-FUNCTIONS/__init_fltk__.cc (plot_window::plot_window):
 	Revert changes from 2011-02-26.
 
 2011-03-10  Konstantinos Poulios  <logari81@googlemail.com>
 
 	* graphics.h.in: Add PERSISTENT mode to listener_mode.
diff --git a/src/graphics.cc b/src/graphics.cc
--- a/src/graphics.cc
+++ b/src/graphics.cc
@@ -4971,41 +4971,31 @@ check_limit_vals (double& min_val, doubl
 
           val = m(2);
           if (! (xisinf (val) || xisnan (val)) && val > 0 && val < min_pos)
             min_pos = val;
         }
     }
 }
 
-// magform(x) Returns (a, b), where x = a * 10^b, a >= 1., and b is
-// integral.
+// magform(x) Returns (a, b), where x = a * 10^b, abs (a) >= 1., and b is
+// integer.
 
 static void
 magform (double x, double& a, int& b)
 {
   if (x == 0)
     {
       a = 0;
       b = 0;
     }
   else
     {
-      double l = std::log10 (std::abs (x));
-      double r = std::fmod (l, 1.);
-      a = std::pow (10.0, r);
-      b = static_cast<int> (l-r);
-      if (a < 1)
-        {
-          a *= 10;
-          b -= 1;
-        }
-
-      if (x < 0)
-        a = -a;
+      b = static_cast<int> (gnulib::floor (std::log10 (std::abs (x)))); 
+      a = x / std::pow (10.0, b); 
     }
 }
 
 // A translation from Tom Holoryd's python code at
 // http://kurage.nimh.nih.gov/tomh/tics.py
 // FIXME -- add log ticks
 
 double
